core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -309,6 +321,104 @@ private void balance() {
+                    break;
+                }
+            }
+            assert currentNode != null : "currently assigned node could not be found";
+
+            // balance the shard, if a better node can be found
+            final float currentWeight = sorter.weight(currentNode);
+            final AllocationDeciders deciders = allocation.deciders();
+            final String idxName = shard.getIndexName();
+            Map<String, NodeRebalanceDecision> nodeDecisions = new HashMap<>();
+            Type rebalanceDecisionType = Type.NO;
+            String assignedNodeId = null;
+            for (ModelNode node : modelNodes) {
+                if (node == currentNode) {
+                    continue; // skip over node we're currently allocated to it
+                }
+                final Decision canAllocate = deciders.canAllocate(shard, node.getRoutingNode(), allocation);
+                final float nodeWeight = sorter.weight(node);
+                final boolean betterWeightThanCurrent = nodeWeight <= currentWeight;
+                final boolean rebalanceConditionsMet;
+                final boolean deltaAboveThreshold;
+                final float weightWithShardAdded;
+                if (betterWeightThanCurrent) {
+                    final float currentDelta = absDelta(nodeWeight, currentWeight);
+                    deltaAboveThreshold = lessThan(currentDelta, threshold) == false;
+                    weightWithShardAdded = weight.weightShardAdded(this, node, idxName);
+                    final float proposedDelta = weightWithShardAdded - weight.weightShardRemoved(this, currentNode, idxName);
+                    rebalanceConditionsMet = deltaAboveThreshold && proposedDelta < currentDelta;
+                    if (rebalanceConditionsMet && canAllocate.type().higherThan(rebalanceDecisionType)) {
+                        // rebalance to the node, only will get overwritten if the decision here is to
+                        // THROTTLE and we get a decision with YES on another node
+                        rebalanceDecisionType = canAllocate.type();
+                        assignedNodeId = node.getNodeId();
+                    }
+                } else {
+                    rebalanceConditionsMet = false;
+                    deltaAboveThreshold = false;
+                    weightWithShardAdded = Float.POSITIVE_INFINITY;
+                }
+                nodeDecisions.put(node.getNodeId(), new NodeRebalanceDecision(
+                    rebalanceConditionsMet ? canAllocate.type() : Type.NO,
+                    canAllocate,
+                    betterWeightThanCurrent,
+                    deltaAboveThreshold,
+                    nodeWeight,
+                    weightWithShardAdded)
+                );
+            }
+
+            return RebalanceDecision.decision(
+                canRebalance, rebalanceDecisionType, assignedNodeId, nodeDecisions, currentWeight, threshold
+            );
```

Comment:
This syntax is a little funky, it feels like bracket syntax, but for a function call.